rewrite-gradle/src/main/java/org/openrewrite/gradle/UpdateGradleWrapper.java

```diff
@@ -127,9 +129,15 @@ private GradleWrapper getGradleWrapper(ExecutionContext ctx) {
                 // services.gradle.org is unreachable, possibly because of a firewall
                 // But if the user specified a wrapperUri to an internal repository things might still be workable
                 if (wrapperUri == null) {
-                    throw new IllegalArgumentException(
-                            "Could not reach services.gradle.org and no alternative wrapper URI is provided. " +
-                            "To use this recipe in environments where services.gradle.org is unavailable specify a wrapperUri.", e);
+                    // If the user didn't specify a wrapperUri, but they did provide a specific version we assume they know this version
+                    // is available from whichever distribution url they were previously using and update the version
+                    if (Semver.validate(version, null).getValue() instanceof ExactVersion) {
+                        return gradleWrapper = new GradleWrapper(version, new DistributionInfos(null, null, null));
```

Comment:
Is there anyway we can go about this change without violating that `null` constraint?

Perhaps we can skip the call to `services.gradle.org` for all exact versions to speed things up as well?

Reply:
The version in this context is being used to select a download and checksum URL from that `services.gradle.org` API call.

To me, it seems like if we have a wrapperUri and the version is an exact version, then we should just forgo calling the `GradleWrapper.create(..)` method.

It's a hard one to trust the existing distribution URL found in the `gradle.properties` file because some organizations add additional version information into the file name. This was why the `wrapperUri` parameter was added because otherwise there is no way to know in a guaranteed manner where should the Gradle version be placed.

Reply:
Unrelated to the above topic, do we potentially want to make the entire distribution info nullable since we aren't carrying any information forward from this point using it?

Reply:
I'd briefly explored that, but it cascades to a lot of then nullable handling we'd have to do in
https://github.com/openrewrite/rewrite/blob/a98d83d698dda9971c93a0767b7f85d5866f306d/rewrite-gradle/src/main/java/org/openrewrite/gradle/util/GradleWrapper.java#L113-L153

Figured the empty string we had before was there for the same reason, even if it's a bit yuck

Reply:
Gotcha.